### PR TITLE
Week14

### PR DESCRIPTION
**Pull request by DYA13**

### CODE REVIEW

**Comment by akosasante** [comment]
Hi @DYA13 , right now it looks like the first commit, committed `node_modules` before adding the `.gitignore` file. So currently the PR is difficult to review since it includes all the thousands of dependency files 😿 . Please follow the steps we've done before to remove the `node_modules`: https://gist.github.com/akosasante/66e3333d13a1858283270d4a35847c4e

**Comment by akosasante** [comment]
Oh also you'll want to include in those same steps, removing `.env` and `.DS_Store`. Removing `.env` is super important, because right now you've exposed your MongoDb URL (which includes the passport to that database cluster!) and session secret to the public 😨 . Please after removing those from git; make sure to also change your mongodb password in Mongo Atlas; and use a different session secret.

**Comment by DYA13** [comment]
> Hi @DYA13 , right now it looks like the first commit, committed `node_modules` before adding the `.gitignore` file. So currently the PR is difficult to review since it includes all the thousands of dependency files 😿 . Please follow the steps we've done before to remove the `node_modules`: https://gist.github.com/akosasante/66e3333d13a1858283270d4a35847c4e


Hello  Akos, actually I added node_modules in . gitignore before , probably some confusion  with git commands. Anyway now it is done as you advised.

**Comment by DYA13** [comment]
> Oh also you'll want to include in those same steps, removing `.env` and `.DS_Store`. Removing `.env` is super important, because right now you've exposed your MongoDb URL (which includes the passport to that database cluster!) and session secret to the public 😨 . Please after removing those from git; make sure to also change your mongodb password in Mongo Atlas; and use a different session secret.

I have changed session secret and mongoDB password.

File: app.js [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+  }
+  res.render("index", { messages });
+});
+
+app.get("/sign-up", (req, res) => {
+  let messages = [];
+  if (req.session.messages) {
+    messages = req.session.messages;
+    req.session.messages = [];
+  }
```
Comment (akosasante) [inline app.js:98]: There's no issue with having this here; but just want to point out that there's currently nothing that is putting messages into the session for the sign up page. 

If you wanted to, you could add messages to the session if errors occur during signup on line 105:

```js
app.post("/sign-up", async (req, res, next) => {
  try {
    const hashedPassword = await bcrypt.hash(req.body.password, 10);
    await User.create({
      username: req.body.username,
      password: hashedPassword,
    });
    res.redirect("/");
  } catch (err) {
    if (!req.session.messages) {
        req.session.messages = [];
    }
    req.session.messages.push(err.message);
    return next(err);
  }
});
```

File: index.ejs [hunk 2]
```diff
@@ -0,0 +1,25 @@
+<!DOCTYPE html>
+<html lang="en">
+  <head>
+    <meta charset="UTF-8" />
+    <title></title>
+  </head>
+  <body>
+    <% if (user) {%>
+    <h1>WELCOME BACK <%= user.username %></h1>
+    <a href="/log-out">LOG OUT</a>
```
Comment (akosasante) [inline index.ejs]: Just an extra suggestion, it could be nice for the user to have a link to the restricted page from here

File: index.ejs [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    <% messages.forEach((msg) =>{ %>
+    <p><%= msg %></p>
+    <% }) %>
+    <form action="/log-in" method="POST">
+      <label for="username">Username</label>
+      <input name="username" placeholder="username" type="text" />
+      <label for="password">Password</label>
+      <input name="password" type="password" />
+      <button>Log In</button>
+    </form>
```
Comment (akosasante) [inline index.ejs]: Just an extra suggestion, it could be nice for the user to have a link to the sign up page from here

File: restricted.ejs [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+  <head>
+    <meta charset="UTF-8" />
+    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
+    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
+    <title>Restricted</title>
+  </head>
+  <body>
+    <p>This page is restricted. You can't see it unless you are logged on.</p>
+    <p>You have visited this page <%= pageCount %> times since logon.</p>
+  </body>
```
Comment (akosasante) [inline restricted.ejs]: Just an extra suggestion, it could be nice for the user to have a link to logout from here

**Review by akosasante — APPROVED** [review]
I left some extra suggestions just if you have more time. Everything looks great and was working well on my end, great job!

After you've cleaned up the `.env` so that it's not showing up in Github, then you should be good to merge 👍

File: index.ejs (hunk 2 above)
Comment (DYA13) [reply index.ejs]: Hi Akos, it is done 👍

File: app.js (hunk 1 above)
Comment (DYA13) [reply app.js:98]: It is done 👍

File: index.ejs (hunk 3 above)
Comment (DYA13) [reply index.ejs]: it is done👍

File: restricted.ejs (hunk 4 above)
Comment (DYA13) [reply restricted.ejs]: it is done👍

**Review by akosasante — APPROVED** [review]
Everything looks good. no longer seeing the .env file and all your additions are working well 👍

**Pull request merged**